mm2src/kdf_walletconnect/src/lib.rs

```diff
@@ -0,0 +1,546 @@
+use metadata::{generate_metadata, AUTH_TOKEN_SUB, PROJECT_ID, RELAY_ADDRESS};
+use mm2_core::mm_ctx::{from_ctx, MmArc};
+use mm2_err_handle::prelude::*;
+use pairing_api::PairingClient;
+use relay_client::websocket::{connection_event_loop as client_event_loop, Client, PublishedMessage};
+use relay_client::{ConnectionOptions, MessageIdGenerator};
+use relay_rpc::auth::{ed25519_dalek::SigningKey, AuthToken};
+use relay_rpc::domain::{MessageId, Topic};
+use relay_rpc::rpc::params::session::Namespace;
+use relay_rpc::rpc::params::session_request::SessionRequestRequest;
+use relay_rpc::rpc::params::{session_request::Request as SessionRequest, IrnMetadata, Metadata, Relay,
+                             RelayProtocolMetadata, RequestParams, ResponseParamsError, ResponseParamsSuccess};
+use relay_rpc::rpc::{ErrorResponse, Payload, Request, Response, SuccessfulResponse};
+use serde::de::DeserializeOwned;
+use session::rpc::delete::send_session_delete_request;
+use session::Session;
+use session::{key::SymKeyPair, SessionManager};
+use std::collections::BTreeSet;
+use std::{sync::Arc, time::Duration};
+use storage::SessionStorageDb;
+use storage::WalletConnectStorageOps;
+use wc_common::{decode_and_decrypt_type0, encrypt_and_encode, EnvelopeType, SymKey};
+
+const PUBLISH_TIMEOUT_SECS: f64 = 6.;
+const MAX_RETRIES: usize = 5;
+
+#[async_trait::async_trait]
+pub trait WalletConnectOps {
+    type Error;
+    type Params<'a>;
+    type SignTxData;
+    type SendTxData;
+
+    async fn wc_chain_id(&self, ctx: &WalletConnectCtx) -> Result<WcChainId, Self::Error>;
+
+    async fn wc_sign_tx<'a>(
+        &self,
+        ctx: &WalletConnectCtx,
+        params: Self::Params<'a>,
+    ) -> Result<Self::SignTxData, Self::Error>;
+
+    async fn wc_send_tx<'a>(
+        &self,
+        ctx: &WalletConnectCtx,
+        params: Self::Params<'a>,
+    ) -> Result<Self::SendTxData, Self::Error>;
+}
+
+pub struct WalletConnectCtx {
+    pub(crate) client: Client,
+    pub(crate) pairing: PairingClient,
+    pub session: SessionManager,
+    pub(crate) key_pair: SymKeyPair,
+    relay: Relay,
+    metadata: Metadata,
+    inbound_message_rx: Arc<Mutex<UnboundedReceiver<PublishedMessage>>>,
+    connection_live_rx: Arc<Mutex<UnboundedReceiver<Option<String>>>>,
+    message_tx: UnboundedSender<SessionMessageType>,
+    message_rx: Arc<Mutex<UnboundedReceiver<SessionMessageType>>>,
+    _abort_handle: AbortOnDropHandle,
+}
+
+impl WalletConnectCtx {
+    pub fn try_init(ctx: &MmArc) -> MmResult<Self, WalletConnectError> {
+        let storage = SessionStorageDb::new(ctx)?;
+        let pairing = PairingClient::new();
+        let relay = Relay {
+            protocol: SUPPORTED_PROTOCOL.to_string(),
+            data: None,
+        };
+        let (inbound_message_tx, inbound_message_rx) = unbounded();
+        let (conn_live_sender, conn_live_receiver) = unbounded();
+        let (message_tx, session_request_receiver) = unbounded();
```

Comment:
let's rename session_request_receiver to message_rx or rename message_tx to session_request_sender.
that aside, it feels wrong holding two ends of the channel in the same struct. you want to have one end given to another party!

Reply:
> it feels wrong holding two ends of the channel in the same struct

why??

Reply:
upd: used weak spawner so this has been refactored.

Reply:
done